### ISSUE TITLE
Upgrade to LLVM 11

### ISSUE THREAD

**Issue by mattmccutchen-cci** [microsoft/checkedc-clang#947]
I believe @kkjeer is in the process of upgrading this repository to LLVM 11.  I'm creating this issue to have a public place to talk about the process that is more appropriate than https://github.com/microsoft/checkedc-clang/pull/930#issuecomment-726152033.

What's the latest plan for how the git history for the merge will be constructed?

Here's what I know so far about the options:

When I first spotted the creation of the `checkedc-clang-110-rc` branch on November 12, it appeared that the plan was to squash all previous Checked C and 3C development into one or a few commits on top of LLVM 11.  We (CCI) objected that this would make it a pain to trace the origin of Checked C or 3C code across the upgrade, and since it is a non-fast-forward change, it would disrupt dependent repositories such as our 5C repository as well.  @mgrang then agreed in an email to @mwhicks1 not to do it this way.

Then yesterday, I saw the `clang-11-upgrade` branch, which seems to be doing it the other way: squashing the difference between our current LLVM 9 baseline and LLVM 11 onto the current Checked C history.  (I took a look and it appears that that the diff that was squashed onto `clang-11-upgrade` may be incorrect, but there may be plans to fix that.)  This approach has an analogous problem of making it a pain to trace the origin of LLVM code across the upgrade, but I don't know if we ever do that, so we likely don't care.  And it is a fast-forward change, so it won't disrupt 5C.  So I think CCI is OK with it.

I wondered why you weren't just doing a normal git merge of the current Checked C with LLVM 11, which is what most people would expect and would keep the history of both sides.  However, I realized there is a potential problem that would occur with the normal merge and not with either of the other approaches.  While the current LLVM 9 baseline (c89a3d78f43d81b9cff7b9248772ddf14d21b749) is on the main line of `llvm/master`, the LLVM 11 commit used in `checkedc-clang-110-rc` (namely, 176249bd6732a8044d457092ed932768724a6f06) is not: it is on `llvm/release/11.x`, which split from `llvm/master` at 2e10b7a39b930ef8d9c4362509d8835b221fbc0a.  Thus, if you later want to upgrade Checked C to a newer commit of LLVM (call it N for this discussion) that is not on `llvm/release/11.x`, and you do it by any fast-forward method, we'll all be stuck with the `llvm/release/11.x` commits in the history, which arguably would be misleading.  This is a significant limitation of git that, as far as I know, has no good solution at this time.  In fact, the most straightforward way to end up with the correct content in the files would be to ours-merge 176249bd6732a8044d457092ed932768724a6f06 into N and then merge the result into the Checked C master.  This could be argued to be a significant enough problem to favor the `clang-11-upgrade` approach on the basis of the end result, regardless of any difference in amount of work between the approaches.

**Comment by kkjeer** [comment]
The `clang-11-upgrade` branch was created using the following approach:

1. Create a pristine `baseline-llvm-110-rc1` branch containing LLVM 11 changes
2. Create a branch `checkedc-clang-110-rc1` branch off of `baseline-llvm-110-rc1`
3. `git merge checkedc-clang/master` onto `checkedc-clang-110-rc1`
4. Create a branch `clang-11-upgrade` off of `checkedc-clang/master`
5. `git merge checkedc-clang-110-rc1` onto `clang-11-upgrade`

However, neither of the `git merge` steps seemed to bring the correct commit history along. I am attempting to debug the reasons why the `git merge` process is not working as expected.

**Comment by mattmccutchen-cci** [comment]
The git history of the new `updated-master` branch looks good to me (I haven't reviewed the merge of the file content) except for one thing: the baseline LLVM commit used, `llvmorg-11-init`, is most similar to LLVM 10.  If you wanted LLVM 11, you should use `llvmorg-12-init` or its parent, just as the current LLVM 9 baseline commit is the parent of `llvmorg-10-init`.  I guess it's your choice whether to use LLVM 10 or 11, though 11 would be convenient for us as we're currently using clang-tidy 11 and it would be nice to build it from our main repository rather than separately.

**Comment by kkjeer** [comment]
Thanks for looking at this @mattmccutchen-cci! @dtarditi should we be using  llvmorg-12-init [Bump the trunk major version to 12](https://github.com/llvm/llvm-project/commit/7ab7b979d29e1e43701cf690f5cf1903740f50e3) or its parent [[LLD][ELF] - Update test after yaml2obj change. NFC.](https://github.com/llvm/llvm-project/commit/2e10b7a39b930ef8d9c4362509d8835b221fbc0a) as our baseline commit?

**Comment by mattmccutchen-cci** [comment]
> @dtarditi should we be using llvmorg-12-init [Bump the trunk major version to 12](https://github.com/llvm/llvm-project/commit/7ab7b979d29e1e43701cf690f5cf1903740f50e3) or its parent [[LLD][ELF] - Update test after yaml2obj change. NFC.](https://github.com/llvm/llvm-project/commit/2e10b7a39b930ef8d9c4362509d8835b221fbc0a) as our baseline commit?

The parent `llvmorg-12-init^` is philosophically more correct and has the practical advantages of showing the correct version number 11 and avoiding merge conflicts if a third party tries to merge Checked C with the LLVM 11 release branch (although we could get those same practical advantages with `llvmorg-12-init` by performing a Checked-C-specific revert of `llvmorg-12-init`).  The only advantage of `llvmorg-12-init` is that when someone comes across it in `git log` output, its significance is immediately clear by both its message and its tag; that isn't true of `llvmorg-12-init^`.  This may not matter if everyone who needs to know about the structure of the Checked C git history is able to find documentation (or this thread).  The LLVM project could have given us the best of both worlds if they made an extra commit and tag just _before_ the release branch splits from master, but we can't expect them to do that if they don't condone what we are doing for the reason described in the next paragraph.

A separate issue: While using `llvmorg-(N+1)-init` (or its parent) rather than the LLVM N release branch dodges git's fundamental "revert a faulty merge" limitation and still allows us to describe our baseline as "similar to LLVM N", is it wise in terms of stability?  My understanding is that the release branch has been heavily tested, while I don't know if `llvmorg-(N+1)-init` can be assumed to be any more stable than LLVM master in general.  If the current `llvmorg-10-init` has been adequately stable, maybe `llvmorg-12-init` will be too.

**Comment by mgrang** [comment]
LLVM tags final releases as `llvmorg-X.Y.Z`. See https://llvm.org/docs/HowToReleaseLLVM.html#tag-the-llvm-final-release. So I think we should upgrade to the `llvmorg-11.0.0` tag.

**Comment by kkjeer** [comment]
I can try to upgrade to [[CodeGen][TailDuplicator] Don't duplicate blocks with INLINEASM_BR](https://github.com/llvm/llvm-project/commit/176249bd6732a8044d457092ed932768724a6f06), which is the latest commit to the `llvmorg-11.0.0` tag.

**Comment by kkjeer** [comment]
The branch `updated-master-11-release` is based on upgrading to the latest commit to `llvmorg-11.0.0` and merging in checkedc-clang changes. I'm working on resolving build failures on that branch.

**Comment by mattmccutchen-cci** [comment]
Cool!  FWIW, I can again confirm that the history looks right to me (I haven't looked at the file content).  So to be clear, you're planning on the ours-merge approach for the next LLVM upgrade?  I think it's probably the least evil (I'm reversing my assessment from the initial comment), but I wanted to confirm that you understand the implications.

We (CCI) believe our understanding with your team was that #930 would be merged before the LLVM 11 upgrade given that there have been no unresolved review issues with #930 since November 23, so the LLVM upgrade branch would need to be updated at that point, although that could likely be done via a normal merge to your master after it has received our PR.  (I haven't looked at how bad the textual and semantic conflicts are: hopefully not bad.  I'll look if you want.)

**Comment by kkjeer** [comment]
I've merged `checkedc-clang/master` into `updated-master-11-release` to reflect the merge of #930 into master and made several more commits to fix build failures. `ninja clang` now builds without errors. However, running `ninja` results in numerous build failures in 3C. @mattmccutchen-cci would someone on the 3C side be available to help look into these?

**Comment by mattmccutchen-cci** [comment]
> would someone on the 3C side be available to help look into these?

Yes, but probably not until Monday or Tuesday.  It's the end of our work day and we need to do something about the [failures on your master](https://github.com/microsoft/checkedc-clang/pull/930#issuecomment-739011181) first.

**Comment by kkjeer** [comment]
> > would someone on the 3C side be available to help look into these?
> 
> Yes, but probably not until Monday or Tuesday. It's the end of our work day and we need to do something about the [failures on your master](https://github.com/microsoft/checkedc-clang/pull/930#issuecomment-739011181) first.

Monday or Tuesday sounds good; thanks for your help!

**Comment by kkjeer** [comment]
I've pushed several commits that fix the remaining build errors (including the 3C build errors). I am now able to run `ninja clang` and `ninja` on the branch `updated-master-11-release`. My next step is to run `ninja check-clang` and `ninja check` and fix test failures.

**Comment by kkjeer** [comment]
After discussion with @dtarditi, I've created new branches `updated-baseline-12-init` and `updated-master-12-init` for the upgrade to clang 11 sources. These branches are based on the LLVM commit https://github.com/llvm/llvm-project/commit/2e10b7a39b930ef8d9c4362509d8835b221fbc0a.

**Comment by mattmccutchen-cci** [comment]
Sounds good.  A nit: It could be misleading to name these branches `*-12-init` when they are actually based on the _parent_ of `llvmorg-12-init` (i.e., `llvmorg-12-init^`).  You could consider names like `*-12-init-parent`, `*-12-p`, or `*-11-last`.

**Comment by kkjeer** [comment]
Thanks @mattmccutchen-cci for the suggestion. I updated the branch names to `*-11-last`.

**Comment by mattmccutchen-cci** [comment]
It looks like the Clang 11 upgrade was merged last night.  Congratulations!  On top of the functionality benefits I imagine this brings to the Checked C compiler and 3C, I've really been looking forward to some fixes/enhancements that Clang 11 brings to the included tools for developing Clang itself that will make our lives more pleasant.  This upgrade will be great for everyone once we work out any remaining kinks.

You can go ahead and close this issue if there isn't any follow-up work that you want to track under it.  I'll get started on merging the upgrade to CCI's repository.

**Comment by mattmccutchen-cci** [comment]
Actually, I noticed one thing: the [note in the readme](https://github.com/microsoft/checkedc-clang#source-code-update) about the current baseline Clang version should be updated.

**Comment by mattmccutchen-cci** [comment]
Do you have the Windows tests passing yet after the upgrade?  We'll need to get them to pass on our upcoming PR to you.  I tried to do a clean build (`ninja check-clang-3c`) on our Windows VM and the whole VM locked up (black screen, nothing happens when I move the mouse, still using ~3 host CPUs).  That was already happening intermittently before the upgrade, so I don't know if it's the intermittent problem or something different about Clang 11.  Eventually we'll probably have to get to the bottom of this intermittent problem if it doesn't go away with a Windows or Visual Studio update.  (Working with Checked C on a Windows VM has been a struggle for us!  If we could get some help from you troubleshooting our configuration, that would be great.)  But in this case, if you haven't gotten the Clang 11 tests to run on Windows yet, there's no point in us trying.

**Comment by mattmccutchen-cci** [comment]
I got our Clang 11 Windows X64 build to complete on the second try, so we're unblocked to start investigating the test failures in our version of 3C on Windows.  Then I changed our build target to remove unnecessary dependencies, which may speed up our builds enough that the system hangs are no longer a major obstacle to us; we'll see.  After that change, I tried a Windows X86 build and it completed on the first try.  So it's clear we are no longer having testing problems that need to be tracked here specifically in relation to Clang 11, though if the system hangs continue to be a problem, we may seek your help another way.

I see you've fixed the readme in #988, so this issue can be closed if you have no other follow-up work left (e.g., failing tests).
